crates/red_knot_python_semantic/mdtest.py

```diff
@@ -0,0 +1,201 @@
+# /// script
+# requires-python = ">=3.11"
+# dependencies = [
+#     "rich",
+#     "watchfiles",
+# ]
+# ///
+
+from __future__ import annotations
+
+import json
+import os
+import subprocess
+from pathlib import Path
+from typing import Literal, Never
+
+from rich.console import Console
+from watchfiles import Change, watch
+
+CRATE_NAME = "red_knot_python_semantic"
+CRATE_ROOT = Path(__file__).resolve().parent
+MDTEST_DIR = CRATE_ROOT / "resources" / "mdtest"
+
+
+class MDTestRunner:
+    mdtest_executable: Path
+    console: Console
+
+    def __init__(self) -> None:
+        self.console = Console()
+
+    def _run_cargo_test(self, message_format: Literal["human", "json"]) -> str:
+        return subprocess.check_output(
+            [
+                "cargo",
+                "test",
+                "--package",
+                CRATE_NAME,
+                "--no-run",
+                "--color=always",
+                "--message-format",
+                message_format,
+            ],
+            cwd=CRATE_ROOT,
+            env=dict(os.environ, CLI_COLOR="1"),
+            stderr=subprocess.STDOUT,
+            text=True,
+        )
+
+    def _recompile_tests(
+        self, status_message: str, message_on_success: bool = True
+    ) -> bool:
+        with self.console.status(status_message):
+            # Run it with 'human' format in case there are errors:
+            try:
+                self._run_cargo_test(message_format="human")
+            except subprocess.CalledProcessError as e:
+                print(e.output)
+                return False
+
+            # Run it again with 'json' format to find the mdtest executable:
+            json_output = self._run_cargo_test(message_format="json")
+
+            if json_output:
+                self._get_executable_path_from_json(json_output)
+
+        if message_on_success:
+            self.console.print("[dim]Tests compiled successfully[/dim]")
+        return True
+
+    def _get_executable_path_from_json(self, json_output: str) -> None:
+        for json_line in json_output.splitlines():
+            try:
+                data = json.loads(json_line)
+            except json.JSONDecodeError:
+                continue
+            if data.get("target", {}).get("name") == "mdtest":
+                self.mdtest_executable = Path(data["executable"])
+                break
+        else:
+            raise RuntimeError(
+                "Could not find mdtest executable after successful compilation"
+            )
+
+    def _run_mdtest(
+        self, arguments: list[str] | None = None, capture_output: bool = False
+    ) -> subprocess.CompletedProcess:
+        arguments = arguments or []
+        return subprocess.run(
+            [self.mdtest_executable, *arguments],
+            cwd=CRATE_ROOT,
+            env=dict(os.environ, CLICOLOR_FORCE="1"),
+            capture_output=capture_output,
+            text=True,
+            check=False,
+        )
```

Comment:
I'm generally not a _massive_ fan of instance attributes that aren't initialized in `__init__`, because they can lead to `AttributeError`s at unexpected moments. I'd prefer to either add an assertion that the attribute has been initialized here (the suggestion below), or (better) have the attribute declared as `Path | None`, initialize it to `None` in `__init__`, and add an assertion in this method that `self.mdtest_executable is not None` before using it. 

I'd also make `capture_output` keyword-only

```suggestion
    def _run_mdtest(
        self, arguments: list[str] | None = None, *, capture_output: bool = False
    ) -> subprocess.CompletedProcess:
        arguments = arguments or []
        assert hasattr(self, "mdtest_executable"), "Must call `self._recompile_tests()` before calling `self._run_mdtest()`"
        return subprocess.run(
            [self.mdtest_executable, *arguments],
            cwd=CRATE_ROOT,
            env=dict(os.environ, CLICOLOR_FORCE="1"),
            capture_output=capture_output,
            text=True,
            check=False,
        )
```